examples/token/main.aleo

```diff
@@ -17,6 +21,76 @@ record token:
     // The token amount.
     amount as u64.private;
 
+/* Approve */
+
+function approve_public:
+    input r0 as address.public; // approver
+    input r1 as address.public; // spender
+    input r2 as u64.public; // amount spender is allowed to withdraw from approver
+
+    finalize r0 r1 r2;
+
+finalize approve_public:
+    input r0 as address.public; // approver
+    input r1 as address.public; // spender
+    input r2 as u64.public; // increase in amount spender is allowed to withdraw from approver
+
+    hash.ped64 r0 into r3; // hash approver into field
+    hash.ped64 r1 into r4; // hash spender into field
+
+    mul r3 r4 into r5; // multiply approver f by spender f
+
+    // if approvals for approver field times spender field exists, the approved amount is increased.
+    // otherwise, the approved allowance is created.
+    increment approvals[r5] by r2;
+
+/* Unapprove */
+
+function unapprove_public:
+    input r0 as address.public; // approver
+    input r1 as address.public; // spender
+    input r2 as u64.public; // amount spender's allowance is decreasing by
+
+    finalize r0 r1 r2;
+
+finalize unapprove_public:
+    input r0 as address.public; // approver
+    input r1 as address.public; // spender
+    input r2 as u64.public; // decrease in amount spender is allowed to withdraw from approver
+
+    hash.ped64 r0 into r3; // hash approver into field
+    hash.ped64 r1 into r4; // hash spender into field
+
+    mul r3 r4 into r5; // multiply approver field by spender field
+
+    // if approvals for approver field times spender field exists, the approved amount is decreased.
+    // otherwise, the approved allowance is created.
+    decrement approvals[r5] by r2;
+
+/* Transfer From */
+
+function transfer_from_public:
+    input r0 as address.public; // from the approver
+    input r1 as address.public; // to the receiver
+    input r2 as u64.public; // amount to transfer
+
+    finalize self.caller r0 r1 r2;
```

Comment:
When another program calls this, will `self.caller` reflect the `program_id` or the signer address in the `package.json`